File: 4.8-bc/py/stage-2-feature-first/app/features/notifications/repository.py
from __future__ import annotations

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from app.features.notifications.model import Notification


class Repository:
    def __init__(self, session: AsyncSession) -> None:
        self._session = session

    async def save(self, n: Notification) -> None:
        await self._session.execute(
            text(
                "INSERT INTO notifications_messages "
                "(id, user_id, channel, payload, sent_at) "
                "VALUES (:id, :user_id, :channel, :payload, :sent_at)"
            ),
            {
                "id": n.id,
                "user_id": n.user_id,
                "channel": n.channel,
                "payload": n.payload,
                "sent_at": n.sent_at,
            },
        )
        await self._session.commit()

File: 4.8-bc/go/stage-2-feature-first/app/notifications/repository.go
package notifications

import (
	"context"

	"github.com/jackc/pgx/v5/pgxpool"
)

type Repository struct {
	db *pgxpool.Pool
}

func NewRepository(db *pgxpool.Pool) *Repository {
	return &Repository{db: db}
}

func (r *Repository) Save(ctx context.Context, n Notification) error {
	_, err := r.db.Exec(ctx,
		`INSERT INTO notifications_messages (id, user_id, channel, payload, sent_at)
		 VALUES ($1, $2, $3, $4, $5)`,
		n.ID, n.UserID, n.Channel, n.Payload, n.SentAt)
	return err
}
